<center>
    <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-DA0101EN-SkillsNetwork/labs/Module%201/images/IDSNlogo.png" width="300" alt="cognitiveclass.ai logo"  />
</center>

# Introduction  Notebook

Estimated time needed: **10** minutes

## Objectives

After completing this lab you will be able to:

*   Acquire data in various ways
*   Obtain insights from data with Pandas library


<h2>Table of Contents</h2>

<div class="alert alert-block alert-info" style="margin-top: 20px">
<ol>
    <li><a href="https://#data_acquisition">Data Acquisition</a>
    <li><a href="https://#basic_insight">Basic Insight of Dataset</a></li>
</ol>

</div>
<hr>


<h1 id="data_acquisition">Data Acquisition</h1>
<p>
There are various formats for a dataset: .csv, .json, .xlsx  etc. The dataset can be stored in different places, on your local machine or sometimes online.<br>

In this section, you will learn how to load a dataset into our Jupyter Notebook.<br>

In our case, the Breast Cancer Dataset is an online source, and it is in a CSV (comma separated value) format. Let's use this dataset as an example to practice data reading.

<ul>
    <li>Data source: <a href="https://www.kaggle.com/datasets/gunesevitan/breast-cancer-metabric">https://www.kaggle.com/datasets/gunesevitan/breast-cancer-metabric</a>.</li>
    <li>Data type: csv</li>
</ul>

The statistical data obtained from <a href="https://www.kaggle.com/datasets/gunesevitan/breast-cancer-metabric">https://www.kaggle.com/datasets/gunesevitan/breast-cancer-metabric</a> under <a href="https://opendatacommons.org/licenses/odbl/1-0/" target="_blank">Database: Open Database, Contents: © Original Authors</a> license.

The Pandas Library is a useful tool that enables us to read various datasets into a dataframe; our Jupyter notebook platforms have a built-in <b>Pandas Library</b> so that all we need to do is import Pandas without installing.
</p>


you are running the lab in your  browser, so we will install the libraries using `piplite`


In [1]:
#you are running the lab in your  browser, so we will install the libraries using ``piplite``
import piplite
import micropip
await piplite.install(['pandas'])
await piplite.install(['matplotlib'])
await piplite.install(['scipy'])
await piplite.install(['seaborn'])
await micropip.install(['ipywidgets'],keep_going=True)
await micropip.install(['tqdm'],keep_going=True)

If you run the lab locally using Anaconda, you can load the correct library and versions by uncommenting the following:


In [2]:
#install specific version of libraries used in  lab
#! mamba install pandas==1.3.3  -y
#! mamba install numpy=1.21.2 -y

In [3]:
# import pandas library
import pandas as pd
import numpy as np

This function will download the dataset into your browser


In [4]:
#This function will download the dataset into your browser

from pyodide.http import pyfetch

async def download(url, filename):
    response = await pyfetch(url)
    if response.status == 200:
        with open(filename, "wb") as f:
            f.write(await response.bytes())

<h2>Read Data</h2>
<p>
We use <code>pandas.read_csv()</code> function to read the csv file. In the brackets, we put the file path along with a quotation mark so that pandas will read the file into a dataframe from that address. The file path can be either an URL or your local file address.<br>
</p>


In [5]:
path = "breast_cancer.csv"

you will need to download the dataset; if you are running locally, please comment out the following


In [6]:
#you will need to download the dataset; if you are running locally, please comment out the following
await download(path, "breast_cancer.csv")
path="breast_cancer.csv"

This dataset was hosted on IBM Cloud object. Click <a href="https://cocl.us/DA101EN_object_storage?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkDA0101ENSkillsNetwork20235326-2021-01-01">HERE</a> for free storage.


In [7]:
# Import pandas library
import pandas as pd

# Read the online file by the URL provides above, and assign it to variable "df"

df = pd.read_csv(path)

After reading the dataset, we can use the <code>dataframe.head(n)</code> method to check the top n rows of the dataframe, where n is an integer. Contrary to <code>dataframe.head(n)</code>, <code>dataframe.tail(n)</code> will show you the bottom n rows of the dataframe.


In [8]:
# show the first 5 rows using dataframe.head() method
print("The first 5 rows of the dataframe") 
df.head(5)

The first 5 rows of the dataframe


Patient ID  Age at Diagnosis Type of Breast Surgery    Cancer Type  \
0    MB-0000             75.65             Mastectomy  Breast Cancer   
1    MB-0002             43.19      Breast Conserving  Breast Cancer   
2    MB-0005             48.87             Mastectomy  Breast Cancer   
3    MB-0006             47.68             Mastectomy  Breast Cancer   
4    MB-0008             76.97             Mastectomy  Breast Cancer   

                        Cancer Type Detailed Cellularity Chemotherapy  \
0           Breast Invasive Ductal Carcinoma         NaN           No   
1           Breast Invasive Ductal Carcinoma        High           No   
2           Breast Invasive Ductal Carcinoma        High          Yes   
3  Breast Mixed Ductal and Lobular Carcinoma    Moderate          Yes   
4  Breast Mixed Ductal and Lobular Carcinoma        High          Yes   

  Pam50 + Claudin-low subtype  Cohort ER status measured by IHC  ...  \
0                 claudin-low     1.0                   Positve  ...   
1                        LumA     1.0                   Positve  ...   
2                        LumB     1.0                   Positve  ...   
3                        LumB     1.0                   Positve  ...   
4                        LumB     1.0                   Positve  ...   

  Overall Survival Status  PR Status Radio Therapy  \
0                  Living   Negative           Yes   
1                  Living   Positive           Yes   
2                Deceased   Positive            No   
3                  Living   Positive           Yes   
4                Deceased   Positive           Yes   

  Relapse Free Status (Months) Relapse Free Status     Sex  \
0                       138.65        Not Recurred  Female   
1                        83.52        Not Recurred  Female   
2                       151.28            Recurred  Female   
3                       162.76        Not Recurred  Female   
4                        18.55            Recurred  Female   

  3-Gene classifier subtype Tumor Size Tumor Stage  Patient's Vital Status  
0                 ER-/HER2-       22.0         2.0                  Living  
1     ER+/HER2- High Prolif       10.0         1.0                  Living  
2                       NaN       15.0         2.0         Died of Disease  
3                       NaN       25.0         2.0                  Living  
4     ER+/HER2- High Prolif       40.0         2.0         Died of Disease  

[5 rows x 34 columns]

<div class="alert alert-danger alertdanger" style="margin-top: 20px">
<h1> Question #1: </h1>
<b>Check the bottom 10 rows of data frame "df".</b>
</div>


In [9]:
# Write your code below and press Shift+Enter to execute


<details><summary>Click here for the solution</summary>

```python
print("The last 10 rows of the dataframe\n")
df.tail(10)
```


<h3>Rename Columns</h3>
<p>
Take a look at our dataset, it already has a header. But the names of some columns do not suit us, so we can rename them.
</p>
<p>
We will use the <code>rename()</code> function.
This method is very useful if we need to change some selected columns, then we need to specify only the information for the columns to be changed.
</p>

In [10]:
# rename headers list
df = df.rename(columns={'ER status measured by IHC': 'ER Status Measured by IHC',
                   'HER2 status measured by SNP6': 'HER2 Status Measured by SNP6',
                   'Lymph nodes examined positive': 'Lymph Nodes Examined Positive',
                   'Nottingham prognostic index': 'Nottingham Prognostic Index',
                   '3-Gene classifier subtype': '3-Gene Classifier Subtype'})
df.head(10)

Patient ID  Age at Diagnosis Type of Breast Surgery    Cancer Type  \
0    MB-0000             75.65             Mastectomy  Breast Cancer   
1    MB-0002             43.19      Breast Conserving  Breast Cancer   
2    MB-0005             48.87             Mastectomy  Breast Cancer   
3    MB-0006             47.68             Mastectomy  Breast Cancer   
4    MB-0008             76.97             Mastectomy  Breast Cancer   
5    MB-0010             78.77             Mastectomy  Breast Cancer   
6    MB-0014             56.45      Breast Conserving  Breast Cancer   
7    MB-0020             70.00             Mastectomy  Breast Cancer   
8    MB-0022             89.08      Breast Conserving  Breast Cancer   
9    MB-0025             76.24                    NaN  Breast Cancer   

                        Cancer Type Detailed Cellularity Chemotherapy  \
0           Breast Invasive Ductal Carcinoma         NaN           No   
1           Breast Invasive Ductal Carcinoma        High           No   
2           Breast Invasive Ductal Carcinoma        High          Yes   
3  Breast Mixed Ductal and Lobular Carcinoma    Moderate          Yes   
4  Breast Mixed Ductal and Lobular Carcinoma        High          Yes   
5           Breast Invasive Ductal Carcinoma    Moderate           No   
6           Breast Invasive Ductal Carcinoma    Moderate          Yes   
7          Breast Invasive Lobular Carcinoma        High          Yes   
8  Breast Mixed Ductal and Lobular Carcinoma    Moderate           No   
9           Breast Invasive Ductal Carcinoma         NaN          NaN   

  Pam50 + Claudin-low subtype  Cohort ER Status Measured by IHC  ...  \
0                 claudin-low     1.0                   Positve  ...   
1                        LumA     1.0                   Positve  ...   
2                        LumB     1.0                   Positve  ...   
3                        LumB     1.0                   Positve  ...   
4                        LumB     1.0                   Positve  ...   
5                        LumB     1.0                   Positve  ...   
6                        LumB     1.0                   Positve  ...   
7                      Normal     1.0                  Negative  ...   
8                 claudin-low     1.0                   Positve  ...   
9                         NaN     1.0                   Positve  ...   

  Overall Survival Status  PR Status Radio Therapy  \
0                  Living   Negative           Yes   
1                  Living   Positive           Yes   
2                Deceased   Positive            No   
3                  Living   Positive           Yes   
4                Deceased   Positive           Yes   
5                Deceased   Positive           Yes   
6                  Living   Positive           Yes   
7                Deceased   Negative           Yes   
8                Deceased   Negative           Yes   
9                     NaN        NaN           NaN   

  Relapse Free Status (Months) Relapse Free Status     Sex  \
0                       138.65        Not Recurred  Female   
1                        83.52        Not Recurred  Female   
2                       151.28            Recurred  Female   
3                       162.76        Not Recurred  Female   
4                        18.55            Recurred  Female   
5                         2.89            Recurred  Female   
6                       162.17        Not Recurred  Female   
7                        11.74            Recurred  Female   
8                        98.22        Not Recurred  Female   
9                       126.32            Recurred  Female   

  3-Gene Classifier Subtype Tumor Size Tumor Stage  Patient's Vital Status  
0                 ER-/HER2-       22.0         2.0                  Living  
1     ER+/HER2- High Prolif       10.0         1.0                  Living  
2                       NaN       15.0         2.0         Died of Disease  
3                       Na

We can drop missing values along the column "3-Gene Classifier Subtype" as follows:

In [11]:
df = df.dropna(subset=["3-Gene Classifier Subtype"], axis=0)
df.head(20)

Patient ID  Age at Diagnosis Type of Breast Surgery    Cancer Type  \
0     MB-0000             75.65             Mastectomy  Breast Cancer   
1     MB-0002             43.19      Breast Conserving  Breast Cancer   
4     MB-0008             76.97             Mastectomy  Breast Cancer   
5     MB-0010             78.77             Mastectomy  Breast Cancer   
7     MB-0020             70.00             Mastectomy  Breast Cancer   
10    MB-0028             86.41      Breast Conserving  Breast Cancer   
11    MB-0035             84.22             Mastectomy  Breast Cancer   
12    MB-0036             85.49             Mastectomy  Breast Cancer   
17    MB-0050             44.64      Breast Conserving  Breast Cancer   
21    MB-0059             75.58             Mastectomy  Breast Cancer   
22    MB-0060             45.43      Breast Conserving  Breast Cancer   
25    MB-0066             61.49      Breast Conserving  Breast Cancer   
35    MB-0100             68.68             Mastectomy  Breast Cancer   
36    MB-0101             46.89             Mastectomy  Breast Cancer   
37    MB-0102             51.38             Mastectomy  Breast Cancer   
38    MB-0106             49.87             Mastectomy  Breast Cancer   
41    MB-0109             82.53                    NaN  Breast Cancer   
43    MB-0111             54.23             Mastectomy  Breast Cancer   
44    MB-0112             83.89             Mastectomy  Breast Cancer   
45    MB-0113             36.96             Mastectomy  Breast Cancer   

                         Cancer Type Detailed Cellularity Chemotherapy  \
0            Breast Invasive Ductal Carcinoma         NaN           No   
1            Breast Invasive Ductal Carcinoma        High           No   
4   Breast Mixed Ductal and Lobular Carcinoma        High          Yes   
5            Breast Invasive Ductal Carcinoma    Moderate           No   
7           Breast Invasive Lobular Carcinoma        High          Yes   
10           Breast Invasive Ductal Carcinoma    Moderate           No   
11          Breast Invasive Lobular Carcinoma        High           No   
12           Breast Invasive Ductal Carcinoma    Moderate           No   
17  Breast Mixed Ductal and Lobular Carcinoma    Moderate          Yes   
21           Breast Invasive Ductal Carcinoma        High           No   
22           Breast Invasive Ductal Carcinoma        High          Yes   
25           Breast Invasive Ductal Carcinoma        High           No   
35           Breast Invasive Ductal Carcinoma         Low          Yes   
36          Breast Invasive Lobular Carcinoma    Moderate           No   
37          Breast Invasive Lobular Carcinoma        High          Yes   
38           Breast Invasive Ductal Carcinoma    Moderate          Yes   
41           Breast Invasive Ductal Carcinoma        High           No   
43           Breast Invasive Ductal Carcinoma        High           No   
44          Breast Invasive Lobular Carcinoma        High           No   
45           Breast Invasive Ductal Carcinoma         Low          Yes   

   Pam50 + Claudin-low subtype  Cohort ER Status Measured by IHC  ...  \
0                  claudin-low     1.0                   Positve  ...   
1                         LumA     1.0                   Positve  ...   
4                         LumB     1.0                   Positve  ...   
5                         LumB     1.0                   Positve  ...   
7                       Normal     1.0                  Negative  ...   
10                        LumB     1.0                   Positve  ...   
11                        Her2     1.0                  Negative  ...   
12                        LumA     1.0                   Positve  ...   
17                      Normal     1.0                   Positve  ...   
21                        LumA     1.0                   Positve  ...   
22                        LumB     1.0                   Positve  ...   
25                        LumB     1.0

Now, we have successfully read the raw dataset and added the correct headers into the dataframe.


 <div class="alert alert-danger alertdanger" style="margin-top: 20px">
<h1> Question #2: </h1>
<b>Find the name of the columns of the dataframe.</b>
</div>


In [12]:
# Write your code below and press Shift+Enter to execute 


<details><summary>Click here for the solution</summary>

```python
print(df.columns)
```

</details>


<h2>Save Dataset</h2>
<p>
Correspondingly, Pandas enables us to save the dataset to csv. By using the <code>dataframe.to_csv()</code> method, you can add the file path and name along with quotation marks in the brackets.
</p>
<p>
For example, if you would save the dataframe <b>df</b> as <b>breast_cancer.csv</b> to your local machine, you may use the syntax below, where <code>index = False</code> means the row names will not be written.
</p>


We can also read and save other file formats. We can use similar functions like **`pd.read_csv()`** and **`df.to_csv()`** for other data formats. The functions are listed in the following table:


<h2>Read/Save Other Data Formats</h2>

| Data Formate |        Read       |            Save |
| ------------ | :---------------: | --------------: |
| csv          |  `pd.read_csv()`  |   `df.to_csv()` |
| json         |  `pd.read_json()` |  `df.to_json()` |
| excel        | `pd.read_excel()` | `df.to_excel()` |
| hdf          |  `pd.read_hdf()`  |   `df.to_hdf()` |
| sql          |  `pd.read_sql()`  |   `df.to_sql()` |
| ...          |        ...        |             ... |


<h1 id="basic_insight">Basic Insight of Dataset</h1>
<p>
After reading data into Pandas dataframe, it is time for us to explore the dataset.<br>

There are several ways to obtain essential insights of the data to help us better understand our dataset.

</p>


<h2>Data Types</h2>
<p>
Data has a variety of types.<br>

The main types stored in Pandas dataframes are <b>object</b>, <b>float</b>, <b>int</b>, <b>bool</b> and <b>datetime64</b>. In order to better learn about each attribute, it is always good for us to know the data type of each column. In Pandas:

</p>


In [13]:
df.dtypes


Patient ID                         object
Age at Diagnosis                  float64
Type of Breast Surgery             object
Cancer Type                        object
Cancer Type Detailed               object
Cellularity                        object
Chemotherapy                       object
Pam50 + Claudin-low subtype        object
Cohort                            float64
ER Status Measured by IHC          object
ER Status                          object
Neoplasm Histologic Grade         float64
HER2 Status Measured by SNP6       object
HER2 Status                        object
Tumor Other Histologic Subtype     object
Hormone Therapy                    object
Inferred Menopausal State          object
Integrative Cluster                object
Primary Tumor Laterality           object
Lymph Nodes Examined Positive     float64
Mutation Count                    float64
Nottingham Prognostic Index       float64
Oncotree Code                      object
Overall Survival (Months)         

A series with the data type of each column is returned.


In [14]:
# check the data type of data frame "df" by .dtypes
print(df.dtypes)

Patient ID                         object
Age at Diagnosis                  float64
Type of Breast Surgery             object
Cancer Type                        object
Cancer Type Detailed               object
Cellularity                        object
Chemotherapy                       object
Pam50 + Claudin-low subtype        object
Cohort                            float64
ER Status Measured by IHC          object
ER Status                          object
Neoplasm Histologic Grade         float64
HER2 Status Measured by SNP6       object
HER2 Status                        object
Tumor Other Histologic Subtype     object
Hormone Therapy                    object
Inferred Menopausal State          object
Integrative Cluster                object
Primary Tumor Laterality           object
Lymph Nodes Examined Positive     float64
Mutation Count                    float64
Nottingham Prognostic Index       float64
Oncotree Code                      object
Overall Survival (Months)         

<p>
As shown above, it is clear to see that the data type of "Patient ID" and "Type of Breast Surgery" are <code>object</code>, "Age at Diagnosis" and "Cohort" is <code>float64</code>, etc.
</p>
<p>
These data types can be changed; we will learn how to accomplish this in a later module.
</p>


<h2>Describe</h2>
If we would like to get a statistical summary of each column e.g. count, column mean value, column standard deviation, etc., we use the describe method:


This method will provide various summary statistics, excluding <code>NaN</code> (Not a Number) values.


In [15]:
df.describe()

Age at Diagnosis       Cohort  Neoplasm Histologic Grade  \
count       1764.000000  1764.000000                1689.000000   
mean          61.161548     2.625850                   2.409118   
std           12.946276     1.231341                   0.649599   
min           21.930000     1.000000                   1.000000   
25%           51.450000     1.000000                   2.000000   
50%           61.880000     3.000000                   2.000000   
75%           70.760000     3.000000                   3.000000   
max           96.290000     5.000000                   3.000000   

       Lymph Nodes Examined Positive  Mutation Count  \
count                    1700.000000     1660.000000   
mean                        1.995882        5.695181   
std                         3.975571        4.139357   
min                         0.000000        1.000000   
25%                         0.000000        3.000000   
50%                         0.000000        5.000000   
75%                         2.000000        7.000000   
max                        41.000000       80.000000   

       Nottingham Prognostic Index  Overall Survival (Months)  \
count                  1764.000000                1764.000000   
mean                      4.010484                 124.837831   
std                       1.161724                  75.039486   
min                       1.000000                   0.000000   
25%                       3.044000                  60.891667   
50%                       4.040000                 115.816667   
75%                       5.040500                 185.308333   
max                       6.360000                 337.033333   

       Relapse Free Status (Months)   Tumor Size  Tumor Stage  
count                   1764.000000  1743.000000  1300.000000  
mean                     109.909354    26.245990     1.732308  
std                       75.333860    15.376386     0.640465  
min                        0.000000     1.000000     0.000000  
25%                       42.167500    17.000000     1.000000  
50%                      100.740000    22.000000     2.000000  
75%                      167.827500    30.000000     2.000000  
max                      326.020000   182.000000     4.000000

<p>
This shows the statistical summary of all numeric-typed (int, float) columns.<br>

For example, the attribute "Age at Diagnosis" has 1764 counts, the mean value of this column is 61.16, the standard deviation is 12.94, the minimum value is 21.93, 25th percentile is 51.45, 50th percentile is 61.88, 75th percentile is 70.76, and the maximum value is 96.29. <br>

However, what if we would also like to check all the columns including those that are of type object? <br><br>

You can add an argument <code>include = "all"</code> inside the bracket. Let's try it again.

</p>


In [16]:
# describe all the columns in "df" 
df.describe(include = "all")

Patient ID  Age at Diagnosis Type of Breast Surgery    Cancer Type  \
count        1764       1764.000000                   1743           1764   
unique       1764               NaN                      2              2   
top       MB-0000               NaN             Mastectomy  Breast Cancer   
freq            1               NaN                   1046           1762   
mean          NaN         61.161548                    NaN            NaN   
std           NaN         12.946276                    NaN            NaN   
min           NaN         21.930000                    NaN            NaN   
25%           NaN         51.450000                    NaN            NaN   
50%           NaN         61.880000                    NaN            NaN   
75%           NaN         70.760000                    NaN            NaN   
max           NaN         96.290000                    NaN            NaN   

                    Cancer Type Detailed Cellularity Chemotherapy  \
count                               1764        1711         1764   
unique                                 8           3            2   
top     Breast Invasive Ductal Carcinoma        High           No   
freq                                1361         863         1402   
mean                                 NaN         NaN          NaN   
std                                  NaN         NaN          NaN   
min                                  NaN         NaN          NaN   
25%                                  NaN         NaN          NaN   
50%                                  NaN         NaN          NaN   
75%                                  NaN         NaN          NaN   
max                                  NaN         NaN          NaN   

       Pam50 + Claudin-low subtype       Cohort ER Status Measured by IHC  \
count                         1764  1764.000000                      1728   
unique                           7          NaN                         2   
top                           LumA          NaN                   Positve   
freq                           649          NaN                      1349   
mean                           NaN     2.625850                       NaN   
std                            NaN     1.231341                       NaN   
min                            NaN     1.000000                       NaN   
25%                            NaN     1.000000                       NaN   
50%                            NaN     3.000000                       NaN   
75%                            NaN     3.000000                       NaN   
max                            NaN     5.000000                       NaN   

        ... Overall Survival Status  PR Status Radio Therapy  \
count   ...                    1764       1764          1764   
unique  ...                       2          2             2   
top     ...                Deceased   Positive           Yes   
freq    ...                    1003        929          1024   
mean    ...                     NaN        NaN           NaN   
std     ...                     NaN        NaN           NaN   
min     ...                     NaN        NaN           NaN   
25%     ...                     NaN        NaN           NaN   
50%     ...                     NaN        NaN           NaN   
75%     ...                     NaN        NaN           NaN   
max     ...                     NaN        NaN           NaN   

       Relapse Free Status (Months) Relapse Free Status     Sex  \
count                   1764.000000                1763    1764   
unique                          NaN                   2       1   
top                             NaN        Not Recurred  Female   
freq                            NaN                1058    1764   
mean                     109.909354                 NaN     NaN   
std                       75.333860                 NaN     NaN   
min                        0.000000                 NaN     NaN   
25%                    

<p>
Now it provides the statistical summary of all the columns, including object-typed attributes.<br>

We can now see how many unique values there, which one is the top value and the frequency of top value in the object-typed columns.<br>

Some values in the table above show as "NaN". This is because those numbers are not available regarding a particular column type.<br>

</p>


<div class="alert alert-danger alertdanger" style="margin-top: 20px">
<h1> Question #3: </h1>

<p>
You can select the columns of a dataframe by indicating the name of each column. For example, you can select the three columns as follows:
</p>
<p>
    <code>dataframe[[' column 1 ',column 2', 'column 3']]</code>
</p>
<p>
Where "column" is the name of the column, you can apply the method  ".describe()" to get the statistics of those columns as follows:
</p>
<p>
    <code>dataframe[[' column 1 ',column 2', 'column 3'] ].describe()</code>
</p>

Apply the  method to ".describe()" to the columns 'Age at Diagnosis' and 'Overall Survival (Months)'.

</div>


In [17]:
# Write your code below and press Shift+Enter to execute 


<details><summary>Click here for the solution</summary>

```python
df[['Age at Diagnosis', 'Overall Survival (Months)']].describe()
```

</details>


<h2>Info</h2>
Another method you can use to check your dataset is:


It provides a concise summary of your DataFrame.

This method prints information about a DataFrame including the index dtype and columns, non-null values and memory usage.


In [18]:
# look at the info of "df"
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1764 entries, 0 to 1984
Data columns (total 34 columns):
 #   Column                          Non-Null Count  Dtype  
---  ------                          --------------  -----  
 0   Patient ID                      1764 non-null   object 
 1   Age at Diagnosis                1764 non-null   float64
 2   Type of Breast Surgery          1743 non-null   object 
 3   Cancer Type                     1764 non-null   object 
 4   Cancer Type Detailed            1764 non-null   object 
 5   Cellularity                     1711 non-null   object 
 6   Chemotherapy                    1764 non-null   object 
 7   Pam50 + Claudin-low subtype     1764 non-null   object 
 8   Cohort                          1764 non-null   float64
 9   ER Status Measured by IHC       1728 non-null   object 
 10  ER Status                       1764 non-null   object 
 11  Neoplasm Histologic Grade       1689 non-null   float64
 12  HER2 Status Measured by SNP6    17

<h1>Excellent! You have just completed the  Introduction Notebook!</h1>


### Thank you for completing this lab!

## Author

<a href="https://www.linkedin.com/in/joseph-s-50398b136/?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkDA0101ENSkillsNetwork20235326-2021-01-01" target="_blank">Joseph Santarcangelo</a>

### Other Contributors

<a href="https://www.linkedin.com/in/mahdi-noorian-58219234/?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkDA0101ENSkillsNetwork20235326-2021-01-01" target="_blank">Mahdi Noorian PhD</a>

Bahare Talayian

Eric Xiao

Steven Dong

Parizad

Hima Vasudevan

<a href="https://www.linkedin.com/in/fiorellawever/?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkDA0101ENSkillsNetwork20235326-2021-01-01" target="_blank">Fiorella Wenver</a>

<a href="https:// https://www.linkedin.com/in/yi-leng-yao-84451275/ " target="_blank" >Yi Yao</a>.

## Change Log

| Date (YYYY-MM-DD) | Version | Changed By | Change Description                                         |
| ----------------- | ------- | ---------- | ---------------------------------------------------------- |
| 2022-08-23        | 2.4     | Malika     | Import micropip added and parameter for ipwidgets and tqdm |
| 2020-10-30        | 2.3     | Lakshmi    | Changed URL of the csv                                     |
| 2020-09-22        | 2.2     | Nayef      | Added replace() method to remove '?'                       |
| 2020-09-09        | 2.1     | Lakshmi    | Made changes in info method of dataframe                   |
| 2020-08-27        | 2.0     | Lavanya    | Moved lab to course repo in GitLab                         |

<hr>

## <h3 align="center"> © IBM Corporation 2020. All rights reserved. <h3/>
